File: 0881-boats-to-save-people/0881-boats-to-save-people.cpp
class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        sort(people.begin(), people.end());
        int boat = 0;
        int left = 0;
        int right = people.size()-1;
        while(left<=right){
            int sum = people[left] + people[right];
            if(sum<=limit){
                boat++;
                left++;
                right--;
            }
            else{
                boat++;
                right--;
            }
        }
        return boat;
    }
};

File: 0881-boats-to-save-people/0881-boats-to-save-people.rs
impl Solution {
    #[allow(dead_code)]
    #[cold]
    pub fn num_rescue_boats(people: Vec<i32>, limit: i32) -> i32 {
        if people.is_empty() {
            return 0;
        }

        let mut people = people.clone();
        people.sort_unstable();

        let (mut i, mut j, mut ans) = (people.len() - 1, 0, 0);
        while i >= j {
            ans += 1;
            if people[i] + people[j] <= limit {
                j += 1;
            }
            if i == 0 {
                break;
            }
            i -= 1;
        }

        ans
    }
}